src/cargo/ops/registry.rs

```diff
@@ -90,7 +91,32 @@ pub struct PublishOpts<'cfg> {
 
 pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
     let specs = opts.to_publish.to_package_id_specs(ws)?;
+    if specs.len() > 1 {
+        match opts.to_publish {
+            Packages::Default => {
+                bail!("must be specified to select a single package to publish. Check `default-members` or using `-p` argument")
+            }
+            Packages::Packages(_) => {
+                bail!("the `-p` argument must be specified to select a single package to publish")
+            }
+            _ => {}
+        }
+    }
+    if Packages::Packages(vec![]) != opts.to_publish && ws.is_virtual() {
```

Comment:
This check seems wrong to me. In the root of a virtual workspace, if I specify `cargo publish -p foo`, then this will error because the packages list has length 1.  However, I think it should not error.

I think the check should probably look something like:

```suggestion
    if opts.to_publish == Packages::Default && ws.is_virtual() {
```